File: examples/data/dp_b.solver.py
from typing import *

INF = 10 ** 18

def solve(n: int, k: int, h: List[int]) -> int:
    dp = [INF for _ in range(n)]
    dp[0] = 0
    for i in range(1, n):
        for j in range(max(0, i - k), i):
            dp[i] = min(dp[i], dp[j] + abs(h[i] - h[j]))
    return dp[n - 1]

def main() -> None:
    n, k = map(int, input().split())
    h = list(map(int, input().split()))
    ans = solve(n, k, h)
    print(ans)

if __name__ == '__main__':
    main()

File: examples/data/dp_b.solver.cpp
#include <climits>
#include <iostream>
#include <vector>
using namespace std;

int64_t solve(int n, int k, const vector<int64_t> &h) {
  vector<int64_t> dp(n, INT64_MAX);
  dp[0] = 0;
  for (int i = 1; i < n; ++i) {
    for (int j = max(0, i - k); j < i; ++j) {
      dp[i] = min(dp[i], dp[j] + abs(h[i] - h[j]));
    }
  }
  return dp[n - 1];
}

int main() {
  int n, k;
  cin >> n >> k;
  vector<int64_t> h(n);
  for (int i = 0; i < n; ++i) {
    cin >> h[i];
  }
  int64_t ans = solve(n, k, h);
  cout << ans << endl;
  return 0;
}
